src/main/cpp/src/NativeParquetJni.cpp

```diff
@@ -0,0 +1,500 @@
+  std::vector<int> schema_num_children;
+  // There are several places where a struct is stored only for a leaf column (like a column chunk)
+  // This holds the gather map for those cases.
+  std::vector<int> chunk_map;
+};
+
+/**
+ * This class will handle processing column pruning for a schema. It is written as a class because
+ * of JNI we are sending the names of the columns as a depth first list, like parquet does internally.
+ */
+class column_pruner {
+public:
+    /**
+     * Create pruining filter from a depth first flattened tree of names and num_children.
+     * The root entry is not included in names or in num_children, but parent_num_children
+     * should hold how many entries there are in it.
+     */
+    column_pruner(const std::vector<std::string> & names, 
+            const std::vector<int> & num_children, 
+            int parent_num_children): children(), s_id(0), c_id(-1) {
+      add_depth_first(names, num_children, parent_num_children);
+    }
+
+    column_pruner(int s_id, int c_id): children(), s_id(s_id), c_id(c_id) {
+    }
+
+    column_pruner(): children(), s_id(0), c_id(-1) {
+    }
+
+    /**
+     * Given a schema from a parquet file create a set of pruning maps to prune columns from the rest of the footer
+     */
+    column_pruning_maps filter_schema(std::vector<parquet::format::SchemaElement> & schema, bool ignore_case) {
+      CUDF_FUNC_RANGE();
+      // The maps are sorted so we can compress the tree...
+      // These are the outputs of the computation
+      std::map<int, int, std::less<int>> chunk_map;
+      std::map<int, int, std::less<int>> schema_map;
+      std::map<int, int, std::less<int>> num_children_map;
+      // Start off with 0 children in the root, will add more as we go
+      schema_map[0] = 0;
+      num_children_map[0] = 0;
+
+      // num_children_stack and tree_stack hold the current state as we walk though schema
+      std::vector<int> num_children_stack;
+      std::vector<column_pruner*> tree_stack;
+      tree_stack.push_back(this);
+      num_children_stack.push_back(schema[0].num_children);
+
+      uint64_t chunk_index = 0;
+      // We are skipping over the first entry in the schema because it is always the root entry, and
+      //  we already processed it
+      for (uint64_t schema_index = 1; schema_index < schema.size(); schema_index++) {
+        // num_children is optional, but is supposed to be set for non-leaf nodes. That said leaf nodes
+        // will have 0 children so we can just default to that.
+        int num_children = 0;
+        if (schema[schema_index].__isset.num_children) {
+          num_children = schema[schema_index].num_children;
+        }
+        std::string name;
+        if (ignore_case) {
+          name = unicode_to_lower(schema[schema_index].name);
+        } else {
+          name = schema[schema_index].name;
+        }
+        column_pruner * found = nullptr;
+        if (tree_stack.back() != nullptr) {
```

Comment:
can we push nulls to `tree_stack`? if not could we replace this with `!tree_stack.empty()`?

Reply:
also maybe consider out a reference to `tree_stack.back()`

Reply:
Yes we can and do push a nullptr to treestack when the item we are looking for is not found. We still need to keep track of the number of children to finish walking the tree so we push in a nullptr.